### PR TITLE
convert empty string to empty collection or array

### PR DESCRIPTION
**Pull request by Mkabaka**

### CODE REVIEW

**Review by alalag1 — CHANGES_REQUESTED** [review]
Make ConverterUtils provide a converter with explicit rules
- convert `null` to `null`
- convert `empty string` type by type()
  - throw an exception if the target type is a primitive type
  - convert to empty `Collection` or `Array`

> plz do not change the impoprts with '*'

**Review by alalag1 — CHANGES_REQUESTED** [review]

File: restlight-jaxrs-provider/src/main/java/esa/restlight/jaxrs/util/JaxrsMappingUtils.java [hunk 1]
```diff
@@ -35,6 +46,10 @@
     private JaxrsMappingUtils() {
     }
 
+    public static boolean hasDefaultValue(Param param) {
+        return param != null && param.getAnnotation(DefaultValue.class) != null;
+    }
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-jaxrs-provider/src/main/java/esa/restlight/jaxrs/util/JaxrsMappingUtils.java:51]: done!

File: restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractNameAndValueArgumentResolver.java [hunk 2]
```diff
@@ -94,14 +94,30 @@ private NameAndValue updateNamedValueInfo(Param param, NameAndValue nav) {
...
-        if (nav.defaultValue != null) {
+        boolean hasDefaultValue = false;
+        if (nav.hasDefaultValue) {
             defaultValue = nav.defaultValue;
+            hasDefaultValue = true;
         } else if (!nav.required && (useObjectDefaultValueIfRequired(param, nav))) {
-            defaultValue = ObjectUtils.defaultValue(param.type());
+            defaultValue = defaultValue(param.type());
+            hasDefaultValue = true;
+        } else if (Optional.class.isAssignableFrom(param.type())) {
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractNameAndValueArgumentResolver.java:104]: done!

File: restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractNameAndValueArgumentResolver.java [hunk 3]
```diff
@@ -94,14 +94,30 @@ private NameAndValue updateNamedValueInfo(Param param, NameAndValue nav) {
...
             defaultValue = ConverterUtils.forceConvertStringValue((String) defaultValue, param.genericType());
+            hasDefaultValue = true;
+        }
+        return new NameAndValue(name, nav.required, defaultValue, hasDefaultValue);
+    }
+
+    private static Object defaultValue(Class<?> type) {
+        if (Optional.class.isAssignableFrom(type)) {
+            return Optional.empty();
         }
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractNameAndValueArgumentResolver.java:119]: done~

File: restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractRequestBodyArgumentResolver.java [hunk 4]
```diff
@@ -62,56 +61,55 @@ public boolean supports(Param param) {
...
             if (contentType == null || MediaType.TEXT_PLAIN.isCompatibleWith(contentType)) {
                 //ignore empty body.
                 if (request.byteBufBody().readableBytes() == 0) {
-                    return checkRequired(null);
+                    return null;
                 }
-                return checkRequired(converter.apply(request.byteBufBody()
-                        .toString(StandardCharsets.UTF_8)));
+                return converter.apply(request.byteBufBody()
+                        .toString(StandardCharsets.UTF_8));
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractRequestBodyArgumentResolver.java:104]: We'll resolve it in https://github.com/esastack/esa-restlight/issues/31

File: restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java [hunk 5]
```diff
@@ -63,35 +64,39 @@ private ConverterUtils() {
...
-        STRING_CONVERTER_MAP.put(Float.class, Float::valueOf);
+        STRING_CONVERTER_MAP.put(Float.class, v -> StringUtils.isEmpty(v) ? null : Float.valueOf(v));
         STRING_CONVERTER_MAP.put(float.class, Float::parseFloat);
 
         STRING_CONVERTER_MAP.put(Void.class, v -> null);
         STRING_CONVERTER_MAP.put(void.class, v -> null);
 
-        STRING_CONVERTER_MAP.put(BigDecimal.class, v -> BigDecimal.valueOf(Double.parseDouble(v)));
+        STRING_CONVERTER_MAP.put(BigDecimal.class, v -> StringUtils.isEmpty(v)
+                ? null : BigDecimal.valueOf(Double.parseDouble(v)));
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java:97]: done!

File: restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractNameAndValueArgumentResolver.java [hunk 6]
```diff
@@ -94,14 +94,30 @@ private NameAndValue updateNamedValueInfo(Param param, NameAndValue nav) {
             }
         }
         Object defaultValue = null;
-        if (nav.defaultValue != null) {
+        boolean hasDefaultValue = false;
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-core/src/main/java/esa/restlight/core/resolver/arg/AbstractNameAndValueArgumentResolver.java:97]: done!

File: restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java [hunk 7]
```diff
@@ -195,12 +196,19 @@ public static String normaliseDefaultValue(String value) {
...
         };
     }
 
+    private static Object null2ObjectConverter(Class<?> requiredType) {
+        if (Optional.class.isAssignableFrom(requiredType)) {
+            return Optional.empty();
+        }
+        return null;
+    }
+
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java:211]: done!

File: restlight-core/src/main/java/esa/restlight/core/resolver/arg/NameAndValue.java [hunk 8]
```diff
@@ -23,16 +23,20 @@
...
+    public NameAndValue(String name, boolean required, Object defaultValue, boolean hasDefaultValue) {
         this.name = name;
         this.required = required;
         this.defaultValue = defaultValue;
+        this.hasDefaultValue = hasDefaultValue;
+    }
+
+    public NameAndValue(String name, boolean required, Object defaultValue) {
+        this(name, required, defaultValue, defaultValue != null);
     }
```
[the remark this answers is not in the document]
Comment (Mkabaka) [reply restlight-core/src/main/java/esa/restlight/core/resolver/arg/NameAndValue.java:37]: This constructor should be retained.

File: restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java [hunk 9]
```diff
@@ -65,36 +65,36 @@ private ConverterUtils() {
...
     static {
         STRING_CONVERTER_MAP.put(Byte.class, v -> StringUtils.isEmpty(v) ? null : Byte.valueOf(v));
-        STRING_CONVERTER_MAP.put(byte.class, Byte::parseByte);
+        STRING_CONVERTER_MAP.put(byte.class, v -> v == null ? null : Byte.parseByte(v));
 
         STRING_CONVERTER_MAP.put(Character.class, v -> StringUtils.isEmpty(v)
-                ? null : v.length() > 0 ? v.charAt(0) : v);
-        STRING_CONVERTER_MAP.put(char.class, v -> v.length() >= 1 ? v.charAt(0) : v);
+                ? null : v.charAt(0));
+        STRING_CONVERTER_MAP.put(char.class, v -> v == null ? null : v.length() >= 1 ? v.charAt(0) : v);
```
Comment (alalag1) [inline restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java:72]: `v.length() > 0`

**Review by alalag1 — APPROVED** [review]

File: restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java (hunk 9 above)
Comment (Mkabaka) [reply restlight-common/src/main/java/esa/restlight/core/util/ConverterUtils.java:72]: done!

**Pull request merged**